compliance/flush.go

```diff
@@ -0,0 +1,261 @@
+}
+
+// FlushFromNonMasterDefaultNI sends a Flush to the server using an old election ID
+// and  validates that the programmed chain of entries are not removed form the
+// default NI using the Get RPC.
+func FlushFromNonMasterDefaultNI(c *fluent.GRIBIClient, wantACK fluent.ProgrammingResult, t testing.TB, _ ...TestOpt) {
+	addFlushEntriesToNI(c, server.DefaultNetworkInstanceName, wantACK, t)
+
+	// addFlushEntriesToNI increments the election ID so to check with the current value,
+	// we need to subtract two to ensure that we are not the current master.
+	curID := electionID.Load() - 2
+
+	ctx := context.Background()
+	c.Start(ctx, t)
+	defer c.Stop(t)
+
+	_, flushErr := c.Flush().
+		WithElectionID(curID, 0).
+		WithAllNetworkInstances().
+		Send()
+	if flushErr == nil {
+		t.Fatalf("did not get expected error from flush, got: %v", flushErr)
+	}
+
+	s, ok := status.FromError(flushErr)
+	if !ok {
+		t.Fatalf("received invalid error from server, got: %v", flushErr)
+	}
+	if got, want := s.Code(), codes.FailedPrecondition; got != want {
+		t.Fatalf("did not get expected error from server, got code: %s (error: %v), want: %s", got, flushErr, want)
+	}
+
+	checkNIHasNEntries(ctx, c, server.DefaultNetworkInstanceName, 3, t)
+}
+
+// FlushOfSpecificNI programs entries into two network-instances, the default and a named
+// L3VRF. It subsequently issues a Flush RPC and ensures that entries that are within the
+// flushed network instance (the default) are removed, but the others are preserved.
+func FlushOfSpecificNI(c *fluent.GRIBIClient, wantACK fluent.ProgrammingResult, t testing.TB, _ ...TestOpt) {
+	// TODO(robjs): we need to initialise the server with >1 network instance.
+	t.Skip()
+
+	vrfName := "TEST-VRF"
+
+	addFlushEntriesToNI(c, server.DefaultNetworkInstanceName, wantACK, t)
+	addFlushEntriesToNI(c, vrfName, wantACK, t)
+
+	// addFlushEntriesToNI increments the election ID so to check with the current value,
+	// we need to subtract one from the current election ID.
+	curID := electionID.Load() - 1
+
+	ctx := context.Background()
+	c.Start(ctx, t)
+	defer c.Stop(t)
+
+	fr, err := c.Flush().
+		WithElectionID(curID, 0).
+		WithNetworkInstance(server.DefaultNetworkInstanceName).
+		Send()
+	switch {
+	case err != nil:
+		t.Fatalf("got unexpected error from flush, got: %v", err)
+	case fr == nil:
+		t.Fatalf("got nil response from flush, got: %v", fr)
+	}
+
+	checkNIHasNEntries(ctx, c, server.DefaultNetworkInstanceName, 0, t)
+	checkNIHasNEntries(ctx, c, vrfName, 3, t)
+}
+
+// FlushOfAllNIs programs entries in two network instances - the default and a
+// named VRF and ensures that entries are removed from both when the Flush specifies
+// that all network instances are to be removed.
+func FlushOfAllNIs(c *fluent.GRIBIClient, wantACK fluent.ProgrammingResult, t testing.TB, _ ...TestOpt) {
+	// TODO(robjs): we need to initialise the server with >1 network instance.
+	t.Skip()
```

Comment:
nit: Makes sense to skip when running the tests locally. But I wonder if it makes sense to make this optional for running from Ondatra where we can ensure the proper TEST-VRF is created before triggering the compliance test?

Reply:
I agree -- what I was going to do was to add a hook into the local compliance tests to inject VRF state into the fake such that we don't need to support a full gNMI Set. This way we can create the state ahead of time for local tests - and other tests can create the state using gNMI.